cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -0,0 +1,191 @@
+            return [defer(op, None) for op in protocols.decompose_once(op)]
+        elif op.classical_controls:
+            controls = []
+            for c in op.classical_controls:
+                if isinstance(c, value.KeyCondition):
+                    if c.key not in measurement_qubits:
+                        raise ValueError(f'Deferred measurement for key={c.key} not found.')
+                    qubits = measurement_qubits[c.key]
+                    if len(qubits) != 1:
+                        # TODO: Multi-qubit conditions require
+                        # https://github.com/quantumlib/Cirq/issues/4512
+                        # Remember to update docstring above once this works.
+                        raise ValueError('Only single qubit conditions are allowed.')
+                    controls.extend(qubits)
+                else:
+                    raise ValueError('Only KeyConditions are allowed.')
+            return op.without_classical_controls().controlled_by(
+                *controls, control_values=[tuple(range(1, q.dimension)) for q in controls]
+            )
+        return op
+
+    circuit = transformer_primitives.map_operations(
+        circuit, defer, raise_if_add_qubits=False
+    ).unfreeze()
+    for k, qubits in measurement_qubits.items():
+        circuit.append(ops.measure(*qubits, key=k))
+    return circuit
+
+
+CIRCUIT_TYPE = TypeVar('CIRCUIT_TYPE', bound='cirq.AbstractCircuit')
+
+
+@transformer_api.transformer
+def dephase_measurements(
+    circuit: CIRCUIT_TYPE, *, context: Optional['cirq.TransformerContext'] = None
+) -> CIRCUIT_TYPE:
+    """Changes all measurements to a dephase operation.
+
+    This transformer is useful when using a density matrix simulator, when
+    wishing to calculate the final density matrix of a circuit and not simulate
+    the measurements themselves.
+
+    Args:
+        circuit: The circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options
+            for transformers.
+    Returns:
+        A copy of the circuit, with dephase operations in place of all
+        measurements.
+    Raises:
+        ValueError: If the circuit contains classical controls. In this case,
+            it is required to change these to quantum controls via
+            `cirq.defer_measurements` first. Since deferral adds ancilla qubits
+            to the circuit, this is not done automatically, to prevent
+            surprises.
+    """
+
+    context1 = context or transformer_api.TransformerContext()
+
+    def dephase(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if any(t in context1.ignore_tags for t in op.tags):
+            return op
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            return ops.KrausChannel.from_channel(ops.phase_damp(1), key=key).on_each(op.qubits)
```

Comment:
We should note that the classical noise (i.e. `invert_mask` or confusion matrix) is being ignored here as well.

Let's discuss in the cirq sync today to see if anyone has concerns with this.

Reply:
Sure, I can't attend today but this is by design. The main goal here is to allow this to replace `ignore_measurement_results` flag in `DensityMatrixSimulator`. As such, what happens to the classical state post-measurement is irrelevant.